src/format/KeePass2Writer.cpp

```diff
@@ -48,12 +50,24 @@ bool KeePass2Writer::writeDatabase(const QString& filename, Database* db)
  * @param db source database
  * @return true on success
  */
+
 bool KeePass2Writer::writeDatabase(QIODevice* device, Database* db) {
     m_error = false;
     m_errorStr.clear();
 
     // determine KDBX3 vs KDBX4
-    if (db->kdf()->uuid() == KeePass2::KDF_AES_KDBX3 && db->publicCustomData().isEmpty()) {
+    bool hasCustomData = !db->publicCustomData().isEmpty() || (db->metadata()->customData() && !db->metadata()->customData()->isEmpty());
```

Comment:
Entries in `Metadata::customData()` do not force `KeePass2::FILE_VERSION_4`. Just entries within `Database::publicCustomData()`, `Entry::customData()` and `Group::customData()` force the new format .